### PR TITLE
285 configure health checks via the service definition

### LINKED ISSUE #285: Configure health checks via the service definition
**Is your feature request related to a problem? Please describe.**
To add health checks we need to add them to the server. If we have different health checks for different service types, we need to create separate starter files for each service.

**Describe the solution you'd like**
We want to be able to configure which health checks is added to which server by simple form of configuration. That leaves us with a simple starter (jitar.ts) file that registers all health checks and add them during startup based on the service configuration.

```ts
const server = await startServer(moduleImporter);
server.registerHealthCheck('database', new DatabaseHealthCheck());
server.registerHealthCheck('storage', new StorageHealthCheck());
// etc
```

And add a health check per service configuration.

```json
{
    "url": "http://node.example.com:3002",
    "node":
    {
        "gateway": "http://gateway.example.com:3000",
        "repository": "http://repository.example.com:2999",
        "healthChecks": [ "database", "storage" ]
    }
}
```

### PR DESCRIPTION
**Pull request by BasMasking**
Fixes #285 

Changes proposed in this pull request:
- register health checks to the server instance
- add health checks based on configuration
- split the build and start of the server in two separate steps

@MaskingTechnology/jitar

### CODE REVIEW

File: packages/server-nodejs/src/JitarServer.ts [hunk 1]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
+
+        this.#options = ServerOptionsReader.read();
+        this.#configuration = RuntimeConfigurationLoader.load(this.#options.config);
     }
 
-    async start(): Promise<void>
+    async build(): Promise<void>
     {
-        console.log(STARTUP_MESSAGE);
+        const runtime = await RuntimeConfigurator.configure(this.#configuration);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:71]:
```suggestion
        this.#runtime = await RuntimeConfigurator.configure(this.#configuration);
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 2]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
 
-        const options = ServerOptionsReader.read();
-        const configuration = RuntimeConfigurationLoader.load(options.config);
-        const runtime = await RuntimeConfigurator.configure(configuration);
+        const logger = LogBuilder.build(this.#options.loglevel);
 
-        const logger = LogBuilder.build(options.loglevel);
+        this.#addControllers(this.#configuration, runtime, logger);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:75]:
```suggestion
        this.#addControllers(this.#configuration, this.#runtime, this.#logger);
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 3]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
 
-        this.#addControllers(configuration, runtime, logger);
+        this.#url = new URL(this.#configuration.url ?? RuntimeDefaults.URL);
 
-        const url = new URL(configuration.url ?? RuntimeDefaults.URL);
+        this.#runtime = runtime;
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:79]:
```suggestion
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 4]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
-        const options = ServerOptionsReader.read();
-        const configuration = RuntimeConfigurationLoader.load(options.config);
-        const runtime = await RuntimeConfigurator.configure(configuration);
+        const logger = LogBuilder.build(this.#options.loglevel);
 
-        const logger = LogBuilder.build(options.loglevel);
+        this.#addControllers(this.#configuration, runtime, logger);
 
-        this.#addControllers(configuration, runtime, logger);
+        this.#url = new URL(this.#configuration.url ?? RuntimeDefaults.URL);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:77]:
```suggestion
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 5]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
         this.#app.use(express.urlencoded({ extended: true }));
 
         this.#app.disable('x-powered-by');
+
+        this.#options = ServerOptionsReader.read();
+        this.#configuration = RuntimeConfigurationLoader.load(this.#options.config);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:66]:
```suggestion
        this.#configuration = RuntimeConfigurationLoader.load(this.#options.config);
        this.#logger = LogBuilder.build(this.#options.loglevel);
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 6]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
-    async start(): Promise<void>
+    async build(): Promise<void>
     {
-        console.log(STARTUP_MESSAGE);
+        const runtime = await RuntimeConfigurator.configure(this.#configuration);
 
-        const options = ServerOptionsReader.read();
-        const configuration = RuntimeConfigurationLoader.load(options.config);
-        const runtime = await RuntimeConfigurator.configure(configuration);
+        const logger = LogBuilder.build(this.#options.loglevel);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:73]:
```suggestion
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 7]
```diff
@@ -51,37 +61,49 @@ export default class JitarServer
...
-        await this.#startServer(url.port);
+    async start(): Promise<void>
+    {
+        console.log(STARTUP_MESSAGE);
 
-        this.#runtime = runtime;
+        if (this.#url === undefined)
+        {
+            throw new Error('URL is undefined');
+        }
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:89]:
```suggestion
        const url = new URL(configuration.url ?? RuntimeDefaults.URL);
```

**Review by petermasking — CHANGES_REQUESTED** [review]

**Comment by BasMasking** [comment]
@petermasking I accepted your suggestions. The file looked weird after pulling it, so had to remove some whitespace. Ready for second round of review.

I also added the healthChecks: ["database"] in the health check example, as it wasn't there.

File: packages/server-nodejs/src/JitarServer.ts [hunk 8]
```diff
@@ -51,37 +61,40 @@ export default class JitarServer
...
-        this.#addControllers(configuration, runtime, logger);
-
-        const url = new URL(configuration.url ?? RuntimeDefaults.URL);
+        this.#addHealthChecks();
 
         await this.#startServer(url.port);
 
-        this.#runtime = runtime;
-
+        const logger = LogBuilder.build(this.#options.loglevel);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:86]:
```suggestion
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 9]
```diff
@@ -51,37 +61,40 @@ export default class JitarServer
...
         logger.info(`Server started and listening at port ${url.port}`);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:87]:
```suggestion
        this.#logger.info(`Server started and listening at port ${url.port}`);
```

File: packages/server-nodejs/src/JitarServer.ts [hunk 10]
```diff
@@ -51,37 +61,40 @@ export default class JitarServer
...
+
+        this.#options = ServerOptionsReader.read();
+        this.#configuration = RuntimeConfigurationLoader.load(this.#options.config);
+        this.#logger = LogBuilder.build(this.#options.loglevel);
+    }
+
+    async build(): Promise<void>
+    {
+        this.#runtime = await RuntimeConfigurator.configure(this.#configuration);
+        this.#addControllers(this.#configuration, this.#runtime, this.#logger);
```
Comment (petermasking) [inline packages/server-nodejs/src/JitarServer.ts:73]: We only pass instance fields, so we can remove them here and pick them up where they are used.

**Review by petermasking — CHANGES_REQUESTED** [review]

**Review by petermasking — APPROVED** [review]

**Pull request merged**